### DISCUSSION TITLE
Allow `AppendFormatted` extension methods

### DISCUSSION THREAD

**Discussion by ltrzesniewski** [dotnet/csharplang#5931] [Language Ideas]
The [improved interpolated strings proposal](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/improved-interpolated-strings.md#append-method-overload-resolution) states the following:

> Note that these rules do not permit extension methods for the `Append...` calls. We could consider enabling that if we choose, but this is analogous to the enumerator pattern, where we allow `GetEnumerator` to be an extension method, but not `Current` or `MoveNext()`.

I'd like this restriction to be lifted.

I think the comparison of `AppendFormatted` to `Current` or `MoveNext` doesn't hold: I'd rather compare `AppendFormatted` to `GetEnumerator` itself: it allows providing an extension entry point to a language feature. `Current` or `MoveNext` are just an implementation detail.

Besides, as many `AppendFormatted` extension methods could be added as needed, while an enumerator implementation requires _exactly_ one `Current` property and one `MoveNext` method.

Also, it _looks like_ the implementation could be a [one-line change](https://github.com/dotnet/roslyn/blob/f78173458a5a0cce4fa2e53a83911abb068f7ac4/src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs#L809) + tests (it currently states `searchExtensionMethodsIfNecessary: false`).

---

You may ask why I'd need this, and I have a concrete use case:

I've added custom interpolated string handlers to our [zero-allocation logging library](https://github.com/Abc-Arbitrage/ZeroLog/tree/v2.0.0-pre2). Given that we are not allowed to allocate in this context, I couldn't add an unconstrained `AppendFormatted<T>(T value)` overload: an object needs to _know_ how to append itself to a message.

While this could be achieved by using interfaces, those fall short in the following situation. Suppose you have the following assemblies:

- `ZeroLog`, which provides the interpolated string handlers, and doesn't reference any other assembly
- `Library`, which provides a `MutableString` class, but doesn't reference `ZeroLog`
- `Application`, which references both `ZeroLog` and `Library`

I'd like to be able to define the following extension method in the `Application` assembly:

```C#
AppendFormatted<THandler>(this THandler handler, MutableString value)
    where THandler : struct, ILogMessageInterpolatedStringHandler
```

This would be able to _combine_ the features of `ZeroLog` and `Library` together to allow using `MutableString` in log messages.

`AppendFormatted` extension methods would be the cleanest way to achieve that, and would allow writing:

```C#
_log.Info($"Something happened: {mutableString}");
```

Currently, I had to find a workaround by adding indirections:

```C#
_log.Info($"Something happened: {mutableString.AsLoggable()}");
```

(`AsLoggable` is an extension method on `MutableString`, which returns an `AppendOperation<MutableString>` that `ZeroLog` understands)

This works (except for `ref struct` types), but is nowhere near as clean as what `AppendFormatted` extension methods would allow.

**Comment by 333fred** [comment]
> I think the comparison of AppendFormatted to Current or MoveNext doesn't hold: I'd rather compare AppendFormatted to GetEnumerator itself: it allows providing an extension entry point to a language feature. Current or MoveNext are just an implementation detail.

I don't agree with this characterization. Neither of these are implementation details, they're integral to the language features themselves. In both cases, we perform lookup based on some external trigger (presence of an interpolated string or a `foreach` loop) to find a type that can be used to perform the operation (the handler or the enumerator). That type is then used as the context for the operation, and the semantics of the operation depend strongly on the implementation details of that type. There is no good place to understand extension methods in this context: what if you don't have a using in scope? The usability problems of this are particularly concerning for me. `AsLoggable` seems like a good way to adapt these libraries in an obvious way that doesn't suffer the same issues.

**Reply by RikkiGibson** [reply]
> There is no good place to understand extension methods in this context: what if you don't have a using in scope? The usability problems of this are particularly concerning for me.

We did ship [extension GetEnumerator()](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/extension-getenumerator.md). Is there a difference here that I'm missing, which makes looking up an extension GetEnumerator in the 'foreach' scenario viable/usable but not an extension AppendFormatted for interpolated strings?

**Reply by 333fred** [reply]
> We did ship [extension GetEnumerator()](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/extension-getenumerator.md).

Yes, for the `GetEnumerator` method. Not for the `MoveNext` method or the `Current` property on the type returned by `GetEnumerator`.. In my analogy, the interpolated string handler type is the return of `GetEnumerator`.

**Reply by RikkiGibson** [reply]
Got it. It doesn't sound like there's anything special the language can do to introduce something analogous here. The Log method you call has a parameter of some handler type, we use it, and that's it.

> This works (except for `ref struct` types), but is nowhere near as clean as what `AppendFormatted` extension methods would allow.

It looks like this would still be a problem even if support were added for the extension method as requested. Right? This can't be handled generically until #1148 happens.

**Reply by ltrzesniewski** [reply]
> I don't agree with this characterization. Neither of these are implementation details, they're integral to the language features themselves.

Sorry, maybe "implementation detail" isn't the best choice of words. What I meant to say is that enumerator types need to have a fixed _shape_ (one `Current` property and one `bool MoveNext() ` method), therefore taking extension methods into account on those types doesn't make much sense.

On the other hand, interpolated string handlers do not require a fixed type _shape_. Any number of `AppendFormatted` overloads could be provided, so I fully expected extension methods to be taken into account on those as well. When I found out it's not the case, it came out as a genuine surprise. 

> There is no good place to understand extension methods in this context: what if you don't have a using in scope?

The same would happen as for extension `GetEnumerator` methods: the `AppendFormatted` extension method wouldn't be in scope and therefore wouldn't be taken into account during binding.

I _really_ see `AppendFormatted` extension methods as analogous to `GetEnumerator` extension methods. 

> `AsLoggable` seems like a good way to adapt these libraries in an obvious way that doesn't suffer the same issues.

`AsLoggable` has several downsides, the most obvious being that it would be required on some types, but not on other ones, which is not friendly for the library users. Error messages don't help much either if you don't know you need to use an adapter for a given type: it would seem the type is not supported.

`AppendFormatted` extension methods don't have this issue, as long as they're in scope (which can easily be ensured in application code). 

Essentially, `AsLoggable` is just a patch. The need for it wrecked the clean design I had in mind. 😕

> It doesn't sound like there's anything special the language can do to introduce something analogous here.

The improved interpolated strings proposal states otherwise, see the quote in the question. 

> The Log method you call has a parameter of some handler type, we use it, and that's it.

The way I see it, using a type includes using the attached extension methods. 🙂

As I said earlier, it came as a surprise to me that a language feature was disabled for this particular case. 

> It looks like this would still be a problem even if support were added for the extension method as requested. Right?

I was referring to this, which works today:

```C#
public static void AppendFormatted(this TheHandler handler, TheRefStruct value)
```

---

Side note: There are 7 interpolated string handler types in ZeroLog, which are identical except for the constructor. Each of those currently has 72 `AppendFormatted` overloads. It would have been nice to be able to implement these as extension methods. I'm not saying that's what I would necessarily do (especially in library code), but having the option to do so would be welcome.
